### ISSUE TITLE
How to map React components to DOM nodes without refs in devtools?

### ISSUE THREAD

**Issue by tajo** [bvaughn/react-devtools-experimental#317]
This is not an issue with react-devtools-experimental (which looks great btw!) but it is somewhat related to creating React devtools so I thought this could be a good place to ask. Sorry if not!

I'm trying to understand how to map React components to DOM nodes without using the official [ref](https://reactjs.org/docs/refs-and-the-dom.html) api. 

I'm working on a Chrome [extension](https://chrome.google.com/webstore/detail/styletron-devtools/bmjdkiphkpkjkicjjgelklhkpbpfiogb) that adds a new pane into the Elements section of Chrome devtools. When you inspect an element on the page, it gives you some additional information about what (CSS in JS in React) styles were applied. Those styles are known only in runtime since they can be dynamic (based on props).

To make this happen, we create a data-structure that maps DOM nodes (their refs) to the styles data. But afaik, the only way how to get access to the underlying DOM node in React is to use the `ref` prop. So we [create a ref](https://github.com/styletron/styletron/blob/master/packages/styletron-react/src/dev-tool.js#L27) for every single styled React component. It (almost) works but it seems pretty hacky. There are various problems like [composing refs](https://github.com/facebook/react/issues/13029) when user is passing their own ref already.. 

**TLDR; Is there some undocumented / dev-only React API that could help us to build this map without adding `ref` to every single component?**

[This](https://github.com/styletron/styletron/blob/master/packages/devtools-extension/devtools_script.js#L7) is how the extension gets the styles data from the element.

**Comment by bvaughn** [comment]
I'd prefer to keep GH issues here focused on DevTools 😄 just so they're easier to manage. That being said...

> TLDR; Is there some undocumented / dev-only React API that could help us to build this map without adding ref to every single component?

There is a private/internal API that React/DevTools use, but we do not suggest using it as internal data structures may change in non-backwards compatible ways with any future release. Because of this, we [intentionally randomize the attribute](https://github.com/facebook/react/blob/4d949d7641490d6df5a96848628bb3fea6cced6d/packages/react-dom/src/client/ReactDOMComponentTree.js#L11-L15). That being said, it can be identified with JS:
https://github.com/bvaughn/react-devtools-experimental/blob/6c8ae706ee22e6a3f69940be0a51cc9763a3a89f/src/backend/views/Overlay.js#L288-L290

But you would be using it at your own risk. 😄
